include/onnxruntime/core/framework/data_types.h

```diff
@@ -636,6 +645,76 @@ class SequenceType : public NonTensorType<CPPType> {
   }
 };
 
+/**
+ * \brief SequenceTensorBase serves as a base type class for
+ *        Tensor sequences. Akin TensorTypeBase.
+ *        Runtime representation is always TensorSeq.
+ */
+class SequenceTensorBase : public DataTypeImpl {
+ public:
+  static MLDataType Type();
+
+  bool IsCompatible(const ONNX_NAMESPACE::TypeProto& type_proto) const override;
+
+  bool IsTensorSequenceType() const override {
+    return true;
+  }
+
+  const SequenceTensorBase* AsSequenceTensorBase() const override {
+    return this;
+  }
+
+  virtual MLDataType GetElementType() const {
+    // should never reach here.
+    ORT_NOT_IMPLEMENTED(__FUNCTION__, " is not implemented");
+  }
+
+  size_t Size() const override;
+
+  DeleteFunc GetDeleteFunc() const override;
+
+  const ONNX_NAMESPACE::TypeProto* GetTypeProto() const override;
+
+  SequenceTensorBase(const SequenceTensorBase&) = delete;
+  SequenceTensorBase& operator=(const SequenceTensorBase&) = delete;
+
+ protected:
+  SequenceTensorBase();
+  ~SequenceTensorBase();
+
+  ONNX_NAMESPACE::TypeProto& mutable_type_proto();
+
+ private:
+  struct Impl;
+  Impl* impl_;
+};
+
+/**
+ * \brief SequenceTensorType. Use to register sequence for non-tensor types.
+ *
+ *  \param CPPRuntime - We always use TensorSeq
+ *
+ *  \param TensorElemType - one of the primitive types
+ *
+ * \details Usage: ORT_REGISTER_SEQ_TENSOR_TYPE()
+ *          The type is required to have value_type defined
+ */
+template <typename TensorElemType>
+class SequenceTensorType : public SequenceTensorBase {
```

Comment:
Can you elaborate why SequenceTensorType should not derive from NonTensorType<TensorSeq> like MapType does? What problem does this solve?

Reply:
Let's just say that deriving from NonTensorType<> does not solve the problem that NonTensorType<> was created to solve. Specifically, it creates a re-usable templated instantiations of functions common for classes that are deriving from NonTensorType<>. But SequenceTensor does not have anything templated with respect to the runtype which is always TensorSeq, that is why it derives from SequenceTensorBase which a) not templated b) has a singletone instance. There is not any other reason to derive from NonTensorType<>. The name is implying that it should be a base for anything that is not a Tensor but it is not true.

---
In reply to: [352969579](https://github.com/microsoft/onnxruntime/pull/2522#discussion_r352969579) [](ancestors = 352969579)

Reply:
Does this not tie ourselves (hard code) to using TensorSeq for SequenceTensorType? If we've a different implementation for tensor sequences tomorrow for efficiency or other reasons, we can't change that now, right?

Reply:
The choice was made (not by me) to tie the implementation to TensorSeq, the same as with Tensor. Should we want a different implementation, we can a) change TensorSeq b) have a different implementation.

---
In reply to: [353962281](https://github.com/microsoft/onnxruntime/pull/2522#discussion_r353962281) [](ancestors = 353962281)